Introduce DTOs

ashleybroughton opened the pull request:
Anyone is welcome to submit PR's to my [fork](https://github.com/ashleybroughton/CoreWiki/tree/dto) for this pull request to keep work together.

We are looking to replace any direct references to entity objects with data transfer objects (DTO's).
If you are unsure what this means or how to do it, look at the `Files changes` tab above to get an idea.

Pages to convert to DTO

* [x] Create
* [x] CreateArticleFromLink
* [x] Delete
* [x] Details (GET)
* [x] Details (post - creating comments)
* [x] Edit
* [x] History
* [x] Search

ashleybroughton committed:
Introduced DTO for All articles page

ashleybroughton committed:
Introduced DTOs for Details page

erwindevreugd commented on CoreWiki/Pages/Details.cshtml.cs:
```diff
@@ -87,24 +114,23 @@ public class DetailsModel : PageModel
-			var author = await _UserManager.FindByIdAsync(Article.AuthorId.ToString());
 			await _commentRepo.CreateComment(comment);
 
+			// IDEA: Make this an extensibility module, we should only be creating a comment here (single responsibility principle)
```
Great idea! #224

erwindevreugd committed:
Introduced DTO for search page

erwindevreugd commented:
What are your thoughts on what the data flow should be?

Repo->DataModel->Dto->PageModel/PageModel->Dto->DataModel->Repo?

or using the dto's as the page model?

erwindevreugd committed:
Implemented suggested changes

ashleybroughton committed:
Introduced DTO for Create page

ashleybroughton committed:
Introduced DTO for Delete page

ashleybroughton committed:
Introduced DTO for Edit page + refactoring

Makes sure topic doesn't conflict with existing article
Redirect back to view page if nothing has changed (Fixes #222)
Discovered #239 during testing

ashleybroughton committed:
nameof doesn't play nicely with model validation

nameof(Article.Topic) returns "Topic"
model validation is expecting "Article.Topic"

ashleybroughton committed:
Introduced DTO for History page

ashleybroughton committed:
Introduced DTO for CreateArticleFromLink page

Discovered #240

ashleybroughton committed:
Introduced DTO for Comment submit

ashleybroughton commented:
Ready for review.  
Found a few bugs during testing which I have created issues for.  
Have confirmed via dev branch that they were not introduced by these changes.

csharpfritz commented:
Thanks so much for this code... we really appreciate it
